### PR TITLE
Define strong uuid-class types for table_id, table_schema_version and query_id

### LINKED ISSUE #11207: Table UUID should be defined as a concrete type
So it won't get confused with other UUID's we use.
For example, schema has both `id()` and `version()` returning `utils::UUID`
while the former is the associated table's uuid and the latter is the schema version uuid.

As a first step, let's wrap the table uuid with a class that will provide explicit conversion
and implement hashing and printing over it to call the respective UUID functions.

Then rename `schema::id()` to ``schema::table_id()`, returning a `replica::table_id`
and similarly, return `replica::table_id` from `replica::database::find_uuid(ks_name, cf_name)`

### PR DESCRIPTION
**Pull request by bhalevy**
We would like to define more distinct types
that are currently defined as aliases to utils::UUID to identify
resources in the system, like table id and schema
version id.

As with counter_id, the motivation is to restrict
the usage of the distinct types so they can be used
(assigned, compared, etc.) only with objects of
the same type.  Using with a generic UUID will
then require explicit conversion, that we want to
expose.

This series starts with cleaning up the idl header definition
by adding support for `import` and `include` statements in the idl-compiler.
These allow the idl header to become self-sufficient
and then remove manually-added includes from source files.
The latter usually need only the top level idl header
and it, in turn, should include other headers if it depends on them.

Then, a UUID_class template was defined as a shared boiler plate
for the various uuid-class.  First, we convert counter_id to use it,
rather than mimicking utils::UUID on its own.

On top of utils::UUID_class<T>, we define table_id,
table_schema_version, and query_id.

Following up on this series, we should define more commonly used
types like: host_id, streaming_plan_id, paxos_ballot_id.

Fixes #11207

### CODE REVIEW

**Comment by bhalevy** [comment]
Cc @Deexie, task_manager::task_id should be based on utils::UUID_class when it becomes available.

**Comment by psarna** [comment]
@bhalevy CI suggests that while making headers self-sufficient, you uncovered a preexisting sstring.hh dependency missing in per partition limiting, which was concealed by the fact that it included serializer.hh, which also provided string.hh

File: idl-compiler.py [hunk 1]
```diff
@@ -74,6 +74,30 @@ def ns_qualified_name(self):
...
+    def __init__(self, name):
+        super().__init__(name)
+
+    def __str__(self):
+        return f"<Import(name={self.name})>"
+
+    def __repr__(self):
+        return self.__str__()
+
+class Include(ASTBase):
```
Comment (denesb) [inline idl-compiler.py:89]: What is the difference between `import` and `include`?

Comment (denesb) [reply idl-compiler.py:89]: I see it down below, but we need to document it somewhere. Speaking of which, this whole IDL is woefully under-documented (or documentation is just very hard to discover). Not this series's fault of course.

File: idl/idl_test.idl.hh [hunk 2]
```diff
@@ -82,3 +84,7 @@ struct const_template_arg_wrapper {
 struct const_template_arg_test_object {
     std::vector<const_template_arg_wrapper<const simple_compound>> first;
 };
+
+struct frozen_object stub [[writable]] {
```
Comment (denesb) [inline idl/idl_test.idl.hh:88]: Maye move this to a separate patch.

File: db/batchlog_manager.cc [hunk 3]
```diff
@@ -31,13 +31,9 @@
 #include "gms/failure_detector.hh"
 #include "gms/gossiper.hh"
 #include "schema_registry.hh"
-#include "idl/uuid.dist.hh"
 #include "idl/frozen_schema.dist.hh"
-#include "serializer_impl.hh"
-#include "serialization_visitors.hh"
-#include "db/schema_tables.hh"
-#include "idl/uuid.dist.impl.hh"
```
Comment (denesb) [inline db/batchlog_manager.cc:39]: Nice. I always hated the obscure error generated by these missing headers and how sensitive it all was to ordering.

**Comment by bhalevy** [comment]
> @bhalevy CI suggests that while making headers self-sufficient, you uncovered a preexisting sstring.hh dependency missing in per partition limiting, which was concealed by the fact that it included serializer.hh, which also provided string.hh

I wonder why I haven't hit it when building locally.
I'll reproduce and fix.

File: db/hints/sync_point.hh [hunk 4]
```diff
@@ -39,5 +39,25 @@ struct sync_point {
...
+struct sync_point_v1 {
+    utils::UUID host_id;
+    uint16_t shard_count;
+
+    // Sync point information for regular mutation hints
+    db::hints::per_manager_sync_point_v1 regular_sp;
+
+    // Sync point information for materialized view hints
+    db::hints::per_manager_sync_point_v1 mv_sp;
+};
```
Comment (denesb) [inline db/hints/sync_point.hh:60]: Where were these before? And why weren't they needed?

File: utils/UUID.hh [hunk 5]
```diff
@@ -206,7 +206,47 @@ inline std::strong_ordering uuid_tri_compare_timeuuid(bytes_view o1, bytes_view
     return res;
 }
 
-}
+template <typename T>
+class UUID_class {
+    int64_t most_sig_bits;
+    int64_t least_sig_bits;
```
Comment (denesb) [inline utils/UUID.hh:212]: Why not store UUID instance?

File: utils/UUID.hh [hunk 6]
```diff
@@ -206,7 +206,47 @@ inline std::strong_ordering uuid_tri_compare_timeuuid(bytes_view o1, bytes_view
...
+public:
+    UUID_class() = default;
```
Comment (denesb) [inline utils/UUID.hh:214]: What is the intended semantics of a default constructed `UUID_class`? Currently it just has two uninitialized fields and will compare equal/less/more to other instances depending on what garbage that memory happened to contain. I think we need a well defined state for a default constructed `UUID_class`.

File: alternator/streams.cc [hunk 7]
```diff
@@ -74,8 +74,8 @@ struct rapidjson::internal::TypeHelper<ValueType, utils::UUID>
     : public from_string_helper<ValueType, utils::UUID>
 {};
 
-static db_clock::time_point as_timepoint(const utils::UUID& uuid) {
```
Comment (denesb) [inline alternator/streams.cc:77]: s/concrete type/strong type/

File: db/hints/manager.cc [hunk 8]
```diff
@@ -506,7 +506,7 @@ future<> manager::end_point_hints_manager::flush_current_hints() noexcept {
 
 class no_column_mapping : public std::out_of_range {
```
Comment (denesb) [inline db/hints/manager.cc:507]: s/concrete type/strong type/

File: schema_fwd.hh [hunk 9]
```diff
@@ -31,3 +31,27 @@ public:
...
+// temporal. For example, schemas with the same set of columns but created at
+// different times should have different versions. This allows nodes to detect
+// if the version they see was already synchronized with or not even if it has
+// the same structure as the past versions.
+//
+// Schema changes merged in any order should result in the same final version.
+//
+// When table_schema_version changes, schema_tables::calculate_schema_digest() should
+// also change when schema mutations are applied.
+class table_schema_version : public utils::UUID_class<table_schema_version> {
```
Comment (denesb) [inline schema_fwd.hh:47]: nice trick to avoid a dummy tag type

File: alternator/executor.cc [hunk 10]
```diff
@@ -3618,7 +3618,7 @@ static future<executor::request_return_type> do_query(service::storage_proxy& pr
         if (schema->clustering_key_size() > 0) {
```
Comment (denesb) [inline alternator/executor.cc:3618]: s/concrete type/strong type/

**Review by denesb — CHANGES_REQUESTED** [review]
Looks good, I think this is a very welcome change.
Had some comments abut the implementation.

File: idl-compiler.py (hunk 1 above)
Comment (bhalevy) [reply idl-compiler.py:89]: `include` just adds a respective `#include` directive in the generated `.dist.hh` header using the same parameter.

`import name` refers to another `name.idl.hh` "module" and adds `#include "name.dist.hh"` and  `#include "name.dist.impl.hh"` to the respective generated header files.

File: idl/idl_test.idl.hh (hunk 2 above)
Comment (bhalevy) [reply idl/idl_test.idl.hh:88]: sure, np

File: db/hints/sync_point.hh (hunk 4 above)
Comment (bhalevy) [reply db/hints/sync_point.hh:60]: oddly enough, https://github.com/scylladb/scylladb/blob/527b34507910ba5606864468e81718f228257450/db/hints/sync_point.cc#L21

included `idl/hinted_handoff.idl.hh` itself.
This worked somehow, but the idl headers aren't really design to be included verbatim by the .cc sources.
The have a hidden assumption that the "real" data structures they represent are defined somewhere else.

With the added support for `include` and `import` statements in idl, it can no longer be included as it was.

I'll separate it to a preliminary patch to clarify what's going on here.

File: utils/UUID.hh (hunk 5 above)
Comment (bhalevy) [reply utils/UUID.hh:212]: Because it is not trivially-default-constructible (it initializes the members to 0)
and so it can't use the serialization templates for trivial types.
To reduce the churn, I just followed the existing `counter_id` approach which is to mimic a UUID and provide conversion methods.

There's a static_assert that checks that these are POD types.

File: utils/UUID.hh (hunk 6 above)
Comment (bhalevy) [reply utils/UUID.hh:214]: Right.  The intention is to keep it as uninitialized so it will be trivially default constructible (See https://en.cppreference.com/w/cpp/types/is_trivial)

File: alternator/streams.cc (hunk 7 above)
Comment (bhalevy) [reply alternator/streams.cc:77]: ok

File: db/hints/manager.cc (hunk 8 above)
Comment (bhalevy) [reply db/hints/manager.cc:507]: ok

File: schema_fwd.hh (hunk 9 above)
Comment (bhalevy) [reply schema_fwd.hh:47]: inspired by `peering_sharded_service` :)

File: alternator/executor.cc (hunk 10 above)
Comment (bhalevy) [reply alternator/executor.cc:3618]: ok

File: utils/UUID.hh (hunk 6 above)
Comment (denesb) [reply utils/UUID.hh:214]: I think you can still initialize members as in `int64_t most_sig_bits = 0`. The constructor is trivial as long as it is not user-defined (i.e. until the compiler can definitely prove that it is not running any user logic).

Comment (denesb) [reply utils/UUID.hh:214]: If that doesn't work we should chose safety over convenience. We can write a templated serializer that works for all UUID_class types easily.

Comment (bhalevy) [reply utils/UUID.hh:214]: > I think you can still initialize members as in int64_t most_sig_bits = 0. The constructor is trivial as long as it is not user-defined (i.e. until the compiler can definitely prove that it is not running any user logic).

That was one of the first things I tried.  The compiler thinks differently than our intuition :)

> If that doesn't work we should chose safety over convenience. We can write a templated serializer that works for all UUID_class types easily.

I'll take a stab at it.

If this works, We should be able to define utils::UUID itself as derived from UUID_class as well.
So all uuid-classes would be symmetrical.

Comment (denesb) [reply utils/UUID.hh:214]: I think we should still keep using composition instead of inheritance, to avoid accidental up-casting.

Comment (bhalevy) [reply utils/UUID.hh:214]: The idea is not to derive UUID_class<T> from UUID, but that rather define UUID itself as a strong class
like `struct UUID : public UUID_class<UUID>`, and that should be immune to accidental type casting to other types derived from `UUID_class` due to the template type tag.

Comment (denesb) [reply utils/UUID.hh:214]: I don't really follow. How will inheriting `UUID` from some type make it a strong type?

**Comment by tgrabiec** [comment]
We already have something like this as `tagged_id` in raft/internal.hh.

**Comment by bhalevy** [comment]
> We already have something like this as `tagged_id` in raft/internal.hh.

Cool, I see now.  They are practically the same.
I'll convert it to to share the generic UUID_class template.

File: utils/UUID.hh (hunk 6 above)
Comment (bhalevy) [reply utils/UUID.hh:214]: never mind.  I tried the common ancestor approach and it is too complicated.
Next version has a UUID_class<T> base template that carries a utils::UUID member.

The culprit requiring trivial types was in bit_cast, but the requirement was too harsh
as proven by https://en.cppreference.com/w/cpp/numeric/bit_cast requires only `std::is_trivially_copyable_v`.

I'm fixing that along the way.

**Comment by tgrabiec** [comment]
What's the advantage of this approach compared to the existing `tagged_id` approach? The latter is concise and non-intrusive to the UUID class.

**Comment by bhalevy** [comment]
> What's the advantage of this approach compared to the existing `tagged_id` approach? The latter is concise and non-intrusive to the UUID class.

The `tagged_id` is good.  I'm preparing v2 based on it so that both raft::internal::tagged_id and the new types will be based on a common `utils::tagged_uuid<Tag>` template.

**Comment by bhalevy** [comment]
In v2(dc46c19b1c)
- rebased + retested
- based on raft::internal::tagged_id as utils::tagged_uuid
- idl_test: add a struct depending on UUID
  - extracted as separate patch
- idl-compiler: add import and include statements
  - added documentation in `docs/dev/IDL.md`
- idl-compiler: include serialization impl and visitors in generated dist.impl.hh files
  - extracted as separate patch
- db/per_partition_rate_limit: tidy up headers self-sufficiency
  - resolves self-sufficiency issue hit in https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1687/
- db: hints: sync_point: do not include idl definition file
  - extracted as separate patch
- Added patches for cleaning up bit_cast header file and overly restrictive is_trivial_v requirement:
  - bit_cast: use std::bit_cast
  - repair: delete unused include of utils/bit_cast.hh
  - utils: bit_cast: require TriviallyCopyable To
  - counters: counter_id delete requirement for triviality
- raft: migrate tagged_id definition to utils::tagged_uuid: provides infrastructure for the new strong types
  - utils: tagged_uuid: bool: reuse uuid::bool operator
  - utils: tagged_uuid: mark functions noexcept
- Following patches are now bases on utils::tagged_uuid and cleaned up:
  - counters: base counter_id of utils::tagged_uuid
  - counters: counter_id: use base class create_random_id
  - utils: tagged_uuid: rename to_uuid() to uuid()
  - schema, everywhere: define and use table_id as a strong type
  - schema, everywhere: define and use table_schema_version as a strong type
  - query-request, everywhere: define and use query_id as a strong type

File: types/query_id.hh [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include "utils/UUID.hh"
+
+using query_id = utils::tagged_uuid<struct query_id_tag>;
```
Comment (denesb) [inline types/query_id.hh:13]: `types/` directory is for storage system type system, not for random utility types we use. I think `query-request.hh` is a better place for this definition.

**Review by denesb — CHANGES_REQUESTED** [review]

File: utils/bit_cast.hh [hunk 12]
```diff
@@ -9,18 +9,19 @@
...
 #include <type_traits>
 
 template <class T> concept Trivial = std::is_trivial_v<T>;
 template <class T> concept TriviallyCopyable = std::is_trivially_copyable_v<T>;
 
-template <Trivial To>
+template <TriviallyCopyable To>
 To read_unaligned(const void* src) {
-    To dst;
+    uint8_t dst[sizeof(To)];
```
[the remark this answers is not in the document]
Comment (bhalevy) [reply utils/bit_cast.hh:22]: In hindsight we can just leave it as `To dst`.
Sine it's trivially copyable he function can just return the value as it did before.

File: schema_fwd.hh [hunk 13]
```diff
@@ -19,3 +21,5 @@ class schema;
 class schema_extension;
 
 using schema_ptr = seastar::lw_shared_ptr<const schema>;
+
+using table_id = utils::tagged_uuid<struct table_id_tag>;
```
[the remark this answers is not in the document]
Comment (bhalevy) [reply schema_fwd.hh:26]: It's bad to redefine the type in two places that may go out of sync.
Maybe we need to change schema_fwd.hh to schema_basic_types.hh.
FWIW, the definition of schema_ptr here is also not a forward declaration.

File: schema_fwd.hh [hunk 14]
```diff
@@ -23,3 +24,21 @@ class schema_extension;
...
+//
+// Schema changes merged in any order should result in the same final version.
+//
+// When table_schema_version changes, schema_tables::calculate_schema_digest() should
+// also change when schema mutations are applied.
+using table_schema_version = utils::tagged_uuid<struct table_schema_version_tag>;
+
+inline table_schema_version reversed(table_schema_version v) noexcept {
+    return table_schema_version(utils::UUID_gen::negate(v.uuid()));
+}
```
[the remark this answers is not in the document]
Comment (bhalevy) [reply schema_fwd.hh:44]: The problem is that schem.hh is too heavy for most use cases that don't need the full schema definition.
Again, changing schem_fwd.hh to schema_basic_types.hh can satisfy both needs.

File: types/query_id.hh (hunk 11 above)
Comment (bhalevy) [reply types/query_id.hh:13]: ok, query-result.hh include query-request.hh so that could work.

**Comment by bhalevy** [comment]
In v3 (6d80b8270):
- rebased + retested
- utils: bit_cast: require TriviallyCopyable To
  - keep `To dst` (do not use `uint8_t dst[]`)
- schema: rename schema_fwd.hh to schema_basic_types.hh
- query-request, everywhere: define and use query_id as a strong type
  - define in query-request.hh (do not introduce types/query_id.hh)

File: utils/bit_cast.hh (hunk 12 above)
[the remark this answers is not in the document]
Comment (bhalevy) [reply utils/bit_cast.hh]: Done in v3

File: schema_fwd.hh (hunk 13 above)
[the remark this answers is not in the document]
Comment (bhalevy) [reply schema_fwd.hh]: Done in v3

File: schema_fwd.hh (hunk 14 above)
[the remark this answers is not in the document]
Comment (bhalevy) [reply schema_fwd.hh]: Done in v3

File: types/query_id.hh (hunk 11 above)
Comment (bhalevy) [reply types/query_id.hh]: Done in v3

File: docs/dev/IDL.md [hunk 15]
```diff
@@ -30,6 +32,17 @@ declaration, similar to C++.
...
+
+Data types that are defined elsewhere can be used in IDL definition files.
+
+To include definitions from C++ headers, use the `include "filename.hh"` statement.
+The idl compiler generates a respective `#include "filename.hh"` directive in the generated `<mod_name>.dist.hh` file.
+
+To import definitions from other IDL definition files, use the `import <other_mod_name>` statement.
+The idl compiler generates a respective `#include "<other_mod_name>.dist.hh"` directive in the generated `<mod_name>.dist.hh` file
+and a `#include "<other_mod_name>.dist.impl.hh"` directive in the generated `<mod_name>.dist.impl.hh` file
+
```
Comment (avikivity) [inline docs/dev/IDL.md:45]: IDL was designed to mimic C++ syntax, so we can switch to C++ reflection when it becomes available (optimistically in C++26). So I'd rather not add new syntax, which conflicts with C++ module import.

Can we make #include detect that it's including idl? For example, #include "xyz.idl.hh" is importing an idl module.

File: utils/UUID.hh [hunk 16]
```diff
@@ -228,4 +226,8 @@ struct hash<utils::UUID> {
         return size_t((hilo >> 32) ^ hilo);
     }
 };
+
+inline std::ostream& operator<<(std::ostream& out, const utils::UUID& uuid) {
+    return out << uuid.to_sstring();
 }
```
Comment (avikivity) [inline utils/UUID.hh:276]: Why make it inline?

File: counters.hh [hunk 17]
```diff
@@ -18,45 +18,17 @@
 class mutation;
 class atomic_cell_or_collection;
 
-class counter_id {
-    int64_t _least_significant;
-    int64_t _most_significant;
+class counter_id : public utils::tagged_uuid<counter_id> {
```
Comment (avikivity) [inline counters.hh]: Why not: using counter_id = utils::tagged_uuid<class something>? Avoids a wrapper class.

File: counters.hh [hunk 18]
```diff
@@ -18,16 +18,7 @@
...
-public:
-    counter_id() = default;
-    explicit counter_id(utils::UUID uuid) noexcept : utils::tagged_uuid<counter_id>(uuid) {}
-
-    // For tests.
-    static counter_id generate_random() {
-        return counter_id(utils::make_random_uuid());
-    }
-};
+using counter_id = utils::tagged_uuid<struct counter_id_tag>;
```
Comment (avikivity) [inline counters.hh:21]: Oh.

File: alternator/serialization.hh [hunk 19]
```diff
@@ -12,7 +12,7 @@
 #include <string_view>
 #include <optional>
 #include "types.hh"
-#include "schema_fwd.hh"
+#include "schema_basic_types.hh"
```
Comment (avikivity) [inline alternator/serialization.hh:15]: I think this patch is just noise. Nobody complained that schema_fwd.hh wasn't 1000% accurate, everyone understood what it meant. Now they have to learn something new for no gain.

File: db/system_keyspace.cc [hunk 20]
```diff
@@ -85,9 +85,9 @@ api::timestamp_type system_keyspace::schema_creation_timestamp() {
 // FIXME: Make automatic by calculating from schema structure.
 static const uint16_t version_sequence_number = 1;
 
-table_schema_version system_keyspace::generate_schema_version(utils::UUID table_id, uint16_t offset) {
+table_schema_version system_keyspace::generate_schema_version(::table_id table_id, uint16_t offset) {
     md5_hasher h;
-    feed_hash(h, table_id);
+    feed_hash(h, table_id.uuid());
```
Comment (avikivity) [inline db/system_keyspace.cc:90]: We should have feed_hash(h, that_template) so this isn't needed.

File: docs/dev/IDL.md (hunk 15 above)
Comment (bhalevy) [reply docs/dev/IDL.md:45]: I don't know how

File: utils/UUID.hh (hunk 16 above)
Comment (bhalevy) [reply utils/UUID.hh:276]: it has to be inline when moved to the header file.
but actually this was done in an earlier version that tried to base utils::UUID itself and the base template, but I think this is no longer needed.

File: counters.hh (hunk 17 above)
Comment (bhalevy) [reply counters.hh]: For `generate_random` below, but this is only a technical step for the patchset.
It is replaced in the next patch with `using ...`

File: alternator/serialization.hh (hunk 19 above)
Comment (bhalevy) [reply alternator/serialization.hh:15]: Botond thought that adding table_id and table_schema_version to it are out of scope.
Therefore the rename serves to increase the scope of the header file.
I don't mind keeping it as is and add the definitions there.

We can define them just in schema.hh but the occasional users will have to include schema.hh that has a lot of bagage.  Maybe that's ok.

However, I do mind defining it the same way in two places, e.g both in schema.hh and schema_fwd.hh.
That's a very bad habbit.

File: idl/uuid.idl.hh [hunk 21]
```diff
@@ -19,3 +19,7 @@ class UUID final {
 class table_id final {
     utils::UUID uuid();
 };
+
+class table_schema_version final {
+    utils::UUID uuid();
+};
```
Comment (avikivity) [inline idl/uuid.idl.hh:26]: table_id and table_schema_version should be defined near their users, not in a generic uuid module. However, it's probably too hairy to untangle so we can live with it for now.

Comment (bhalevy) [reply idl/uuid.idl.hh:26]: yes, it is used in multiple places and even with the support for 'import' it is hard to decide who's the lowest level user for it so it can be defined there.

**Comment by avikivity** [comment]
Looks good, left some minor comments. 

I think the tagged_uuid explicit constructor is not explicit enough, it's easy to call it by mistake. I suggest to change it to from_uuid(), but not in this series, it's long enough.

File: db/system_keyspace.cc (hunk 20 above)
Comment (bhalevy) [reply db/system_keyspace.cc:90]: Yes.  This does the trick:
```
diff --git a/utils/UUID.hh b/utils/UUID.hh
index 1d86249135..8635c07a5f 100644
--- a/utils/UUID.hh
+++ b/utils/UUID.hh
@@ -254,6 +254,14 @@ struct appending_hash<utils::UUID> {
     }
 };
 
+template<typename Tag>
+struct appending_hash<utils::tagged_uuid<Tag>> {
+    template<typename Hasher>
+    void operator()(Hasher& h, const utils::tagged_uuid<Tag>& id) const noexcept {
+        appending_hash<utils::UUID>{}(h, id.uuid());
+    }
+};
+
 namespace std {
 template<>
 struct hash<utils::UUID> {
```

File: idl/uuid.idl.hh (hunk 21 above)
Comment (avikivity) [reply idl/uuid.idl.hh:26]: The standard approach is to define their own modules. But let's not do it now.

File: utils/UUID.hh (hunk 16 above)
Comment (avikivity) [reply utils/UUID.hh:276]: Why can't you have a declaration but not a definition?

File: docs/dev/IDL.md (hunk 15 above)
Comment (avikivity) [reply docs/dev/IDL.md:45]: str.endswith(".idl.hh")?

File: idl/uuid.idl.hh (hunk 21 above)
Comment (bhalevy) [reply idl/uuid.idl.hh:26]: yes, I thought of `schema.idl.hh`, but we I agree it's an overkill for just those 2 basic types.

File: utils/UUID.hh (hunk 16 above)
Comment (bhalevy) [reply utils/UUID.hh:276]: As I tried to explain above, I can.
It was a preparation patch from an older version that is no longer needed.

File: docs/dev/IDL.md (hunk 15 above)
Comment (bhalevy) [reply docs/dev/IDL.md:45]: ah, so it will look like C pre-processor, but will do more.
that's possible.

File: alternator/serialization.hh (hunk 19 above)
Comment (avikivity) [reply alternator/serialization.hh:15]: Doesn't schema.hh include schema_fwd.hh?

Comment (bhalevy) [reply alternator/serialization.hh:15]: It definitely should have, but it didn't up until this series.

**Comment by bhalevy** [comment]
In v4 (eed8124f4c):
- idl-compiler: add include statements
  - use `#include` for both cases
  - updated doc respectively
- dropped: "utils: uuid: define operator<< inline in header file"
- added "utils: uuid: define appending_hash<utils::tagged_uuid<Tag>>"
- added "schema: include schema_fwd.hh in schema.hh"
- dropped "schema: rename schema_fwd.hh to schema_basic_types.hh"
  - use schem_fwd.hh to define both `table_id` and `table_schema_version` using `utils::tagged_uuid`

**Review by avikivity — APPROVED** [review]

**Review by denesb — APPROVED** [review]
I'm not happy about `schema_fwd.hh` becoming a second `schema.hh`, housing a random subset of the schema-related types. In my world, a `*_fwd.hh` header is not a file I ever have to open to look for a definition: it just contains short re-declarations of stuff already declared fully elsewhere. Yes, in some cases there is no shorter form of declarations, so some stuff may be repeated.
But it looks like others don't mind this, furthermore this practice already has precedents, so I won't object.

**Comment by denesb** [comment]
Conflicts, please rebase.

**Comment by bhalevy** [comment]
c71ef330b2 rebased on top of ba42852b0e1d2bec677e60a863c505a637884e04

**Comment by annastuchlik** [comment]
@scylladb/scylla-maint Please backport to 5.1

**Comment by bhalevy** [comment]
> @scylladb/scylla-maint Please backport to 5.1

why?

**Comment by denesb** [comment]
We are preparing for per-version documentation.

**Comment by bhalevy** [comment]
> We are preparing for per-version documentation.

But this isn't a documentation change

**Comment by denesb** [comment]
This might be a mistake then, @annastuchlik ?
